### PR TITLE
Error handling in ERC20 and ERC721

### LINKED ISSUE #1701: Refactoring function ownerOf in ERC721.sol
Check the presence of a token using the function  _exists.  Now the ownerOf function corresponds to the getApproved function.

```
function ownerOf(uint256 tokenId) public view returns (address) {
        require(_exists(tokenId));
        return _tokenOwner[tokenId];
}
```

### PR DESCRIPTION
**Pull request by princesinha19**
I did some error handling in ERC20 and ERC721. In every require, I added some messages so that developer or end user can easily understand what's wrong.

Fixes #1701

### CODE REVIEW

**Comment by princesinha19** [comment]
@frangio @nventuro Please have a look at this PR. I added message in require statement and also made some changes in ERC20.sol and ERC721.sol. Please let me know If you guys think some changes are need to be done.

**Comment by nventuro** [comment]
Hello @princesinha19, thank you for contributing!

We've received multiple requests for this (ever since #888) and will be working on it shortly: see my related comment [here](https://github.com/OpenZeppelin/openzeppelin-solidity/pull/1704#issuecomment-478840315) for more relevant information.

**Comment by princesinha19** [comment]
okay cools @nventuro

**Comment by nventuro** [comment]
Please see [this post in our forum](https://forum.zeppelin.solutions/t/openzeppelin-solidity-requires-do-not-have-string-reason-messages/388) for discussion regarding revert reasons, and share your thoughts!

**Comment by nventuro** [comment]
@princesinha19 I opened https://github.com/OpenZeppelin/openzeppelin-solidity/issues/1709 with some guidelines regarding message format, among other things, including some examples. Could you review it and update your PR accordingly? Thanks!

File: contracts/token/ERC20/ERC20.sol [hunk 1]
```diff
@@ -125,7 +125,7 @@ contract ERC20 is IERC20 {
      * @param value The amount to be transferred.
      */
     function _transfer(address from, address to, uint256 value) internal {
-        require(to != address(0));
+        require(to != address(0), "empty or zero transaction recipient");
```
Comment (nventuro) [inline contracts/token/ERC20/ERC20.sol]: I'd change this to `transfer` instead of `transaction`, which may cause the reader to think we're talking about the Ethereum transaction.

I'd also go with 'zero' as opposed to 'empty' when talking about the zero address.

File: contracts/token/ERC20/ERC20.sol [hunk 2]
```diff
@@ -140,7 +140,7 @@ contract ERC20 is IERC20 {
      * @param value The amount that will be created.
      */
     function _mint(address account, uint256 value) internal {
-        require(account != address(0));
+        require(account != address(0), "empty or zero transaction recipient");
```
Comment (nventuro) [inline contracts/token/ERC20/ERC20.sol]: Similarly, this should be `mint`, and the others `burn`, `approve`, etc. No two reason strings should read the same.

File: contracts/token/ERC721/ERC721.sol [hunk 3]
```diff
@@ -57,7 +57,7 @@ contract ERC721 is ERC165, IERC721 {
      * @return uint256 representing the amount owned by the passed address
      */
     function balanceOf(address owner) public view returns (uint256) {
-        require(owner != address(0));
+        require(owner != address(0), "empty or zero transaction recipient");
```
Comment (nventuro) [inline contracts/token/ERC721/ERC721.sol]: There's no transaction here, just a query. Maybe something like `balance query for zero address`?

File: contracts/token/ERC721/ERC721.sol [hunk 4]
```diff
@@ -68,7 +68,7 @@ contract ERC721 is ERC165, IERC721 {
      */
     function ownerOf(uint256 tokenId) public view returns (address) {
         address owner = _tokenOwner[tokenId];
-        require(owner != address(0));
+        require(owner != address(0), "empty or zero transaction recipient");
```
Comment (nventuro) [inline contracts/token/ERC721/ERC721.sol]: This `require` actually means that there is no owner.

File: contracts/token/ERC721/ERC721.sol [hunk 5]
```diff
@@ -81,9 +81,10 @@ contract ERC721 is ERC165, IERC721 {
      * @param tokenId uint256 ID of the token to be approved
      */
     function approve(address to, uint256 tokenId) public {
+        require(_exists(tokenId), "tokenId to approve doesn't exist");
```
Comment (nventuro) [inline contracts/token/ERC721/ERC721.sol]: What is the reason for this extra `require`? `ownerOf` already asserts this.

File: contracts/token/ERC721/ERC721.sol [hunk 6]
```diff
@@ -148,7 +149,7 @@ contract ERC721 is ERC165, IERC721 {
      * @param tokenId uint256 ID of the token to be transferred
      */
     function safeTransferFrom(address from, address to, uint256 tokenId) public {
-        safeTransferFrom(from, to, tokenId, "");
+        safeTransferFrom(from, to, tokenId);
```
Comment (nventuro) [inline contracts/token/ERC721/ERC721.sol]: What is the reason behind this change? This makes it be an infinite loop.

File: contracts/token/ERC20/ERC20.sol (hunk 1 above)
Comment (princesinha19) [reply contracts/token/ERC20/ERC20.sol]: sorry, my mistake. It should be transfer. I will change is to "zero transfer recipient".

File: contracts/token/ERC20/ERC20.sol (hunk 2 above)
Comment (princesinha19) [reply contracts/token/ERC20/ERC20.sol]: yes correct.

File: contracts/token/ERC721/ERC721.sol (hunk 3 above)
Comment (princesinha19) [reply contracts/token/ERC721/ERC721.sol]: I will change it.

File: contracts/token/ERC721/ERC721.sol (hunk 6 above)
Comment (princesinha19) [reply contracts/token/ERC721/ERC721.sol]: Sorry, I forgot to remove this.

File: contracts/token/ERC721/ERC721.sol (hunk 5 above)
Comment (princesinha19) [reply contracts/token/ERC721/ERC721.sol]: Yes right. I will remove it.

**Comment by princesinha19** [comment]
thanks, @nventuro for your suggestions.
I made the changes as per your suggestions. Please have a look.

File: contracts/token/ERC20/ERC20.sol (hunk 1 above)
Comment (princesinha19) [reply contracts/token/ERC20/ERC20.sol]: I changed it to "transfer recipient is zero". I think it's more meaningful. Please suggest if something I need to change.

**Comment by nventuro** [comment]
@princesinha19 to keep things organized, I merged this PR into https://github.com/OpenZeppelin/openzeppelin-solidity/pull/1704 (see [this](https://github.com/OpenZeppelin/openzeppelin-solidity/pull/1704/commits/13df19db6b42d69d29500a5b62a7c7e309abe353) commit), while doing some slight modifications to a couple messages, so I'm closing this PR now in favor of the other one.

Thank you very much for your contribution!

**Comment by princesinha19** [comment]
Okay, no problem. @nventuro
